gradle-plugins/compose/src/test/kotlin/org/jetbrains/compose/test/tests/integration/DesktopApplicationTest.kt

```diff
@@ -571,6 +574,103 @@ class DesktopApplicationTest : GradlePluginTestBase() {
         }
     }
 
+    @Suppress("SameParameterValue")
+    private fun appCdsProject(
+        appCdsMode: AppCdsMode,
+        javaVersion: Int,
+        javaVendor: JvmVendor.KnownJvmVendor = JvmVendor.KnownJvmVendor.AMAZON
+    ) : TestProject {
+        return testProject("application/appCds").apply {
+            modifyText("build.gradle.kts") {
+                it
+                    .replace("%APP_CDS_MODE%", "AppCdsMode.$appCdsMode")
+                    .replace("%JAVA_VERSION%", "$javaVersion")
+                    .replace("%JVM_VENDOR%", javaVendor.name)
+            }
+        }
+    }
+
+    @Test
+    fun testAppCdsAutoFailsOnJdk17() = with(appCdsProject(AppCdsMode.Auto, 17)) {
+        fun testRunTask(runTask: String) {
+            gradleFailure(runTask).checks {
+                check.logContains("AppCdsMode 'Auto' is not supported on JDK earlier than 19; current is 17")
+            }
+        }
+
+        testRunTask(":runDistributable")
+    }
+
+    private val loggedArchivePathRegex =
+        Regex("\\[cds] Opened archive " +
+                listOf(
+                    ".*",
+                    "build",
+                    "compose",
+                    "binaries",
+                    ".*",
+                    "app",
+                    ".*",
+                    AppCdsMode.ARCHIVE_FILE_ARGUMENT
+                        .replace("\$APPDIR", "app")
+                        .replace(".", "\\.")
+                ).joinToString(File.separator.replace("\\", "\\\\"))
+        )
+
+    @Test
+    fun testAppCdsAuto() = with(appCdsProject(AppCdsMode.Auto, 21)) {
+        fun testRunTask(taskName: String) {
+            gradle(taskName).checks {
+                check.taskSuccessful(taskName)
+                check.logContains("[cds] Dumping shared data to file")
+            }
+            gradle(taskName).checks {
+                check.taskSuccessful(taskName)
+                check.logContainsMatch(loggedArchivePathRegex)
+                check.logDoesntContain("[cds] Specified shared archive not found")
+                check.logDoesntContain("[cds] Dumping shared data to file")
+                check.logDoesntContain("[cds] Initialize dynamic archive failed")
+                check.logDoesntContain("[cds] An error has occurred while processing the shared archive file")
+                check.logDoesntContain("[cds] Failed to initialize dynamic archive")
+            }
+            gradle(":clean")
+        }
+
+        testRunTask(":runDistributable")
+        testRunTask(":runReleaseDistributable")
+    }
```

Comment:
I don't know the context, but why do we run each task twice?
and why is the second run not up-to-date?

Reply:
> I don't know the context, but why do we run each task twice?

In `Auto` mode, the first time the app runs it creates the AppCDS archive. Then the following runs use the archive. The test checks that this is indeed what happens.

> and why is the second run not up-to-date?

These are `run` tasks... They always run.